src/coreclr/vm/syncblk.cpp

```diff
@@ -2541,15 +2545,19 @@ BOOL AwareLock::EnterEpilogHelper(Thread* pCurThread, INT32 timeOut)
     // the object associated with this lock.
     _ASSERTE(pCurThread->PreemptiveGCDisabled());
 
-    BOOLEAN IsContentionKeywordEnabled = ETW_TRACING_CATEGORY_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, TRACE_LEVEL_INFORMATION, CLR_CONTENTION_KEYWORD);
+    bool isContentionKeywordEnabled = IsContentionKeywordEnabled();
     LARGE_INTEGER startTicks = { {0} };
 
-    if (IsContentionKeywordEnabled)
+    if (isContentionKeywordEnabled)
     {
         QueryPerformanceCounter(&startTicks);
 
         // Fire a contention start event for a managed contention
-        FireEtwContentionStart_V1(ETW::ContentionLog::ContentionStructs::ManagedContention, GetClrInstanceId());
+        FireEtwContentionStart_V2(
+            ETW::ContentionLog::ContentionStructs::ManagedContention,
+            GetClrInstanceId(),
+            this,
+            m_HoldingThread);
```

Comment:
The last argument here should actually be the OS thread ID.  This allows users of the event to match the lock up with the thread that is holding it in a trace.  The memory address of the lock won't allow for this.

Reply:
I commented on this in https://github.com/dotnet/runtime/pull/72627#issuecomment-1224119479. It looks like the holding thread pointer cannot be dereferenced here because there's not guarantee that it would be alive. The `ThreadRunning` event is emitted from a newly started thread and can be used to map the `Thread` pointer to the OS thread ID.

Reply:
For threads that are created from outside the runtime and then enter the runtime, I believe the `ThreadCreated` event can be used to do the same mapping.

Reply:
I don't think we can depend on the `ThreadRunning` event because there is nothing that guarantees that this event is present in the trace.  If the thread already exists when the trace is started, then we have no way to correlate it back to the OS thread.  That said, this call is made in a function that requires cooperative mode.  Given this, even if the thread dies, the memory shouldn't be reclaimed until after the next GC when the managed object Finalizer executes, right?

Reply:
> I don't think we can depend on the ThreadRunning event because there is nothing that guarantees that this event is present in the trace. If the thread already exists when the trace is started, then we have no way to correlate it back to the OS thread.

That's also true for the new `LockCreated` event. It seems like the ordering of those events would matter, so having a rundown of thread and lock objects may not help much to identify the thread. Perhaps the thread object is stored as the holding thread because it is easily accessible and is the minimum work necessary to identify the holding thread. Storing the OS thread ID instead would involve an indirection, and OS thread IDs may also be reused at some point, I suspect this kind of issue would exist in multiple places.

> That said, this call is made in a function that requires cooperative mode. Given this, even if the thread dies, the memory shouldn't be reclaimed until after the next GC when the managed object Finalizer executes, right?

I'm not sure about that, coop GC mode would ensure that the managed thread object would be alive if accessible, but once the managed thread object is deemed unreachable and the finalizer runs, perhaps the native thread object can be deleted irrespective of coop GC mode. Maybe there are further constraints that I'm not aware of?

Reply:
Perhaps something that would fill the gap is to have an event that is emitted upon starting to collect a trace (not rundown), which maps the native thread object to an OS thread ID. Would that be feasible?

An alternative may be to ensure that it is ok to dereference the holding thread field somehow when emitting the event.

Reply:
We could just check if the `m_holdingThread` is null before dereferencing it to avoid crashing. Pass 0 if it's null, the OS Thread ID instead.

Reply:
What's preventing the `Thread` object from being deleted after reading a non-null value for the field?

Reply:
How would it be an issue  ?
The thread id would be informational and if someone wants to use (stackwalking for example), they will have to check if the thread is dead or not.

Reply:
If the thread is dead, reading the ID from the Thread* can crash the process - the memory for Thread* can be unmapped.

Reply:
Oh ok, I was thinking @kouvel was referring to an invalid OS Thread ID, which is not a problem per se.
But yeah, accessing unmapped memory would be bad.

Thanks @jkotas for the clarification.

Reply:
Ordering of `LockCreated` and `Contention` is important - if you don't have the `LockCreated` event, then it's possible that if the ID is re-used, that you won't know it.  It doesn't make the data useless, but could cause someone to conflate the existence of two locks down to one.  At the same time, emitting the `LockCreated` event more often than we should (e.g. every time there is a call to `GetSyncBlock`) would yield the opposite problem - we could see many creation events for a single lock creation.

With regard to the OS thread ID.  You're right that these can be re-used.  This is addressed because the OS will tell us when a thread is created/destroyed.  Thus, if there is a re-use within a trace, we'd see this from the OS thread events.

I was trying to convince myself that the managed thread object would solve the liveness issue, but I think you're right that it does not, since it's possible for there to not be a managed `Thread` object.  A way to address this would be for the thread that acquires the lock to save its thread ID into the lock when tracing is on and clear it when the lock is relinquished.  What do you think of that?  There are cases where this wouldn't work (e.g. the thread acquires the lock and blocks indefinitely before the trace is started), but I'm not too concerned about this, and still like the idea of keeping the overhead low, such that we don't save and clear the OS thread ID when tracing is disabled.

Reply:
> emitting the LockCreated event more often than we should (e.g. every time there is a call to GetSyncBlock)

Why would the `LockCreated` event be emitted every time `GetSyncBlock()` is called?

> With regard to the OS thread ID. You're right that these can be re-used. This is addressed because the OS will tell us when a thread is created/destroyed. Thus, if there is a re-use within a trace, we'd see this from the OS thread events.

Why would there not be a similar problem to what you mentioned before regarding missing events when starting a trace on an already-running process? For instance, the create/destroy OS events for threads may not be captured by the trace.

> A way to address this would be for the thread that acquires the lock to save its thread ID into the lock when tracing is on and clear it when the lock is relinquished. What do you think of that?

The information stored is probably minimal, in the sense what is stored is what is most easily accessible for correct behavior, a different implementation may store something else and that should be ok. As I mentioned before, storing the thread ID would involve an indirection. Changing the implementation for this particular reason is probably something that would need a lot more scrutiny, I don't see enough reason for it yet.

Reply:
> Why would the LockCreated event be emitted every time GetSyncBlock() is called?

You're right, it shouldn't.  I missed the fact that the cache lookup occurs at the beginning of this function, and returns early.

> Why would there not be a similar problem to what you mentioned before regarding missing events when starting a trace on an already-running process? For instance, the create/destroy OS events for threads may not be captured by the trace.

You technically don't need to see the thread creation event to know that it existed at the beginning of the trace, as long as you see samples or events on the thread.  What you do need is to have a thread creation event during the trace to know that the thread ID has been re-used.  That said, the OS provides thread/process rundown events at the beginning of the trace to tell profilers about the existence of all processes and traces.

> The information stored is probably minimal, in the sense what is stored is what is most easily accessible for correct behavior, a different implementation may store something else and that should be ok. As I mentioned before, storing the thread ID would involve an indirection. Changing the implementation for this particular reason is probably something that would need a lot more scrutiny, I don't see enough reason for it yet.

I'm not sure I follow.  Can you tell me a little bit more here?  Are you saying that you don't think having the OS Thread ID is worthwhile here?

Reply:
> That said, the OS provides thread/process rundown events at the beginning of the trace to tell profilers about the existence of all processes and traces.

Curious, does this work with circular buffers? More generally, is there some mechanism by which a trace taken with a circular buffer, where the buffer fills up and is overwritten, can still have an accurate snapshot of live objects through events at the beginning of the trace?

> Are you saying that you don't think having the OS Thread ID is worthwhile here?

No, mapping to a thread ID would probably be the most useful part. I don't think we should change or add to what is stored for the lock-holding thread. Perhaps taking the thread store lock could enable emitting the OS thread ID.

Reply:
A thread that orphans a lock does not clear the holding thread pointer before its `Thread` object is destroyed, so acquiring the thread store lock still would not make it feasible to dereference the holding thread pointer. There seem to be a few other related issues and fixing them properly would likely be more involved. For now, we could also store the OS thread ID when the event keyword is enabled as you suggested and update it after the other issues are fixed.

Reply:
Updated in the latest commit to store and send the OS thread ID, since I think it would be possible to continue doing so after the complete fixes. Still curious about the question in https://github.com/dotnet/runtime/pull/72627#discussion_r955313883, as I've run into that issue before and a solution to it may be useful elsewhere.